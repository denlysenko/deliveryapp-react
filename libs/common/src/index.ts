export * from './lib/constants';
export * from './lib/errors';
export * from './lib/roles';
export * from './lib/app-menu';
export * from './lib/messages';
export * from './lib/payment-methods';
