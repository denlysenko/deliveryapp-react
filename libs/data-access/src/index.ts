import * as authClient from './lib/api/auth';
import * as usersClient from './lib/api/users';
import * as ordersClient from './lib/api/orders';
import * as paymentsClient from './lib/api/payments';

export { authClient, usersClient, ordersClient, paymentsClient };

export * from './lib/models/auth';
export * from './lib/models/user';
export * from './lib/models/validation-error';
export * from './lib/models/order';
export * from './lib/models/list-response';
export * from './lib/models/orders-filter';
export * from './lib/models/payment';
export * from './lib/models/payments-filter';
export * from './lib/models/users-filter';

export * from './lib/context/auth';
export * from './lib/context/orders';
export * from './lib/context/payments';
