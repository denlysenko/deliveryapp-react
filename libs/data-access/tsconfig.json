{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "jsx": "react",
    "esModuleInterop": true,
    "types": ["node", "jest"]
  },
  "include": ["**/*.ts"]
}
