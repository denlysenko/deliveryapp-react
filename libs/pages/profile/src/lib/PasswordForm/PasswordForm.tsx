import React, { useState } from 'react';

import { useFormik } from 'formik';
import * as Yup from 'yup';

import { InputText } from 'primereact/inputtext';
import { Button } from 'primereact/button';
import { Message } from 'primereact/message';

import { ERRORS, handleValidationError } from '@deliveryapp/common';
import { updatePassword } from '@deliveryapp/data-access';

import { StyledForm } from '../StyledForm';

const ValidationSchema = Yup.object().shape({
  oldPassword: Yup.string().required(ERRORS.REQUIRED_FIELD),
  newPassword: Yup.string().required(ERRORS.REQUIRED_FIELD),
  confirmPassword: Yup.string()
    .oneOf([Yup.ref('newPassword')], ERRORS.PASSWORDS_DO_NOT_MATCH)
    .required(ERRORS.REQUIRED_FIELD)
});

export const PasswordForm: React.FC<{}> = () => {
  const [loading, setLoading] = useState(false);

  const formik = useFormik({
    initialValues: {
      oldPassword: '',
      newPassword: '',
      confirmPassword: ''
    },
    validationSchema: ValidationSchema,
    onSubmit: async values => {
      setLoading(true);

      try {
        const { oldPassword, newPassword } = values;
        await updatePassword({ oldPassword, newPassword });
        setLoading(false);
      } catch (error) {
        setLoading(false);
        handleValidationError(error.response.data, formik);
      }
    }
  });

  return (
    <StyledForm>
      <h3>Change Password</h3>
      <div className="p-grid">
        <div className="p-col-12 row">
          <div className="input-wrapper p-float-label">
            <InputText
              id="oldPassword"
              type="password"
              data-testid="oldPassword"
              className={
                formik.touched.oldPassword && formik.errors.oldPassword
                  ? 'invalid'
                  : ''
              }
              value={formik.values.oldPassword}
              onChange={formik.handleChange}
            />
            <label htmlFor="oldPassword">Current Password</label>
            <i className="fa fa-lock"></i>
          </div>
          {formik.touched.oldPassword && formik.errors.oldPassword && (
            <Message
              id="oldPassword-error"
              severity="error"
              text={formik.errors.oldPassword}
            ></Message>
          )}
        </div>
        <div className="p-col-12 row">
          <div className="input-wrapper p-float-label">
            <InputText
              id="newPassword"
              type="password"
              data-testid="newPassword"
              className={
                formik.touched.newPassword && formik.errors.newPassword
                  ? 'invalid'
                  : ''
              }
              value={formik.values.newPassword}
              onChange={formik.handleChange}
            />
            <label>New Password</label>
            <i className="fa fa-lock"></i>
          </div>
          {formik.touched.newPassword && formik.errors.newPassword && (
            <Message
              id="newPassword-error"
              severity="error"
              text={formik.errors.newPassword}
            ></Message>
          )}
        </div>
        <div className="p-col-12 row">
          <div className="input-wrapper p-float-label">
            <InputText
              id="confirmPassword"
              type="password"
              data-testid="confirmPassword"
              className={
                formik.touched.confirmPassword && formik.errors.confirmPassword
                  ? 'invalid'
                  : ''
              }
              value={formik.values.confirmPassword}
              onChange={formik.handleChange}
            />
            <label htmlFor="confirmPassword">Confirm Password</label>
            <i className="fa fa-lock"></i>
          </div>
          {formik.touched.confirmPassword && formik.errors.confirmPassword && (
            <Message
              id="confirmPassword-error"
              severity="error"
              text={formik.errors.confirmPassword}
            ></Message>
          )}
        </div>
        <div className="p-col-12 row center">
          <Button
            label="Change Password"
            type="button"
            data-testid="change-password"
            className="p-button-raised"
            disabled={loading}
            onClick={formik.submitForm}
          />
        </div>
      </div>
    </StyledForm>
  );
};
