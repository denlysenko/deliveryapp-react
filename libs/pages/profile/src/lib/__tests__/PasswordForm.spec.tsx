import React from 'react';
import { render, fireEvent, cleanup, wait } from '@testing-library/react';

import { ERRORS } from '@deliveryapp/common';
import { usersClient } from '@deliveryapp/data-access';

import { PasswordForm } from '../PasswordForm/PasswordForm';

const oldPassword = 'oldPassword';
const newPassword = 'newPassword';

describe('[Profile] PasswordForm', () => {
  beforeEach(() => {
    jest
      .spyOn(usersClient, 'updatePassword')
      .mockImplementation(() => Promise.resolve());
  });

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  it('should render password form', () => {
    const { baseElement } = render(<PasswordForm />);
    expect(baseElement).toMatchSnapshot();
  });

  describe('Validations', () => {
    describe('oldPassword', () => {
      it('should display required error', async () => {
        const { getByTestId, container } = render(<PasswordForm />);

        await wait(() => {
          fireEvent.click(getByTestId('change-password'));
        });

        expect(container.querySelector('#oldPassword-error')).toHaveTextContent(
          ERRORS.REQUIRED_FIELD
        );
      });

      it('should not display any errors', async () => {
        const { getByTestId, container } = render(<PasswordForm />);

        fireEvent.change(getByTestId('oldPassword'), {
          target: {
            value: oldPassword
          }
        });

        await wait(() => {
          fireEvent.click(getByTestId('change-password'));
        });

        expect(
          container.querySelector('#oldPassword-error')
        ).not.toBeInTheDocument();
      });
    });

    describe('newPassword', () => {
      it('should display required error', async () => {
        const { getByTestId, container } = render(<PasswordForm />);

        await wait(() => {
          fireEvent.click(getByTestId('change-password'));
        });

        expect(container.querySelector('#newPassword-error')).toHaveTextContent(
          ERRORS.REQUIRED_FIELD
        );
      });

      it('should not display any errors', async () => {
        const { getByTestId, container } = render(<PasswordForm />);

        fireEvent.change(getByTestId('newPassword'), {
          target: {
            value: newPassword
          }
        });

        await wait(() => {
          fireEvent.click(getByTestId('change-password'));
        });

        expect(
          container.querySelector('#newPassword-error')
        ).not.toBeInTheDocument();
      });
    });

    describe('confirmPassword', () => {
      it('should display required error', async () => {
        const { getByTestId, container } = render(<PasswordForm />);

        await wait(() => {
          fireEvent.click(getByTestId('change-password'));
        });

        expect(
          container.querySelector('#confirmPassword-error')
        ).toHaveTextContent(ERRORS.REQUIRED_FIELD);
      });

      it('should display passwords do not match error', async () => {
        const { getByTestId, container } = render(<PasswordForm />);

        fireEvent.change(getByTestId('newPassword'), {
          target: {
            value: newPassword
          }
        });

        fireEvent.change(getByTestId('confirmPassword'), {
          target: {
            value: 'wrong'
          }
        });

        await wait(() => {
          fireEvent.click(getByTestId('change-password'));
        });

        expect(
          container.querySelector('#confirmPassword-error')
        ).toHaveTextContent(ERRORS.PASSWORDS_DO_NOT_MATCH);
      });

      it('should not display any errors', async () => {
        const { getByTestId, container } = render(<PasswordForm />);

        fireEvent.change(getByTestId('newPassword'), {
          target: {
            value: newPassword
          }
        });

        fireEvent.change(getByTestId('confirmPassword'), {
          target: {
            value: newPassword
          }
        });

        await wait(() => {
          fireEvent.click(getByTestId('change-password'));
        });

        expect(
          container.querySelector('#confirmPassword-error')
        ).not.toBeInTheDocument();
      });
    });
  });

  describe('Submitting', () => {
    it('should be successful', async () => {
      const { getByTestId } = render(<PasswordForm />);

      fireEvent.change(getByTestId('oldPassword'), {
        target: {
          value: oldPassword
        }
      });

      fireEvent.change(getByTestId('newPassword'), {
        target: {
          value: newPassword
        }
      });

      fireEvent.change(getByTestId('confirmPassword'), {
        target: {
          value: newPassword
        }
      });

      await wait(() => {
        fireEvent.click(getByTestId('change-password'));
      });

      expect(usersClient.updatePassword).toBeCalledTimes(1);
      expect(usersClient.updatePassword).toBeCalledWith({
        oldPassword,
        newPassword
      });
    });

    it('should have error', async () => {
      const error = {
        errors: [{ path: 'oldPassword', message: 'Error' }]
      };

      jest.spyOn(usersClient, 'updatePassword').mockImplementationOnce(() =>
        Promise.reject({
          response: {
            data: error
          }
        })
      );

      const { getByTestId, container } = render(<PasswordForm />);

      fireEvent.change(getByTestId('oldPassword'), {
        target: {
          value: oldPassword
        }
      });

      fireEvent.change(getByTestId('newPassword'), {
        target: {
          value: newPassword
        }
      });

      fireEvent.change(getByTestId('confirmPassword'), {
        target: {
          value: newPassword
        }
      });

      await wait(() => {
        fireEvent.click(getByTestId('change-password'));
      });

      expect(container.querySelector('#oldPassword-error')).toHaveTextContent(
        error.errors[0].message
      );
    });
  });
});
