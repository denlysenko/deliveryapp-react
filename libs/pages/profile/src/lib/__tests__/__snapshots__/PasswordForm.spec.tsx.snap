// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Profile] PasswordForm should render password form 1`] = `
<body>
  <div>
    <div
      class="sc-AxjAm icXODU"
    >
      <h3>
        Change Password
      </h3>
      <div
        class="p-grid"
      >
        <div
          class="p-col-12 row"
        >
          <div
            class="input-wrapper p-float-label"
          >
            <input
              class="p-inputtext p-component"
              data-testid="oldPassword"
              id="oldPassword"
              name="oldPassword"
              type="password"
              value=""
            />
            <label
              for="oldPassword"
            >
              Current Password
            </label>
            <i
              class="fa fa-lock"
            />
          </div>
        </div>
        <div
          class="p-col-12 row"
        >
          <div
            class="input-wrapper p-float-label"
          >
            <input
              class="p-inputtext p-component"
              data-testid="newPassword"
              id="newPassword"
              name="newPassword"
              type="password"
              value=""
            />
            <label>
              New Password
            </label>
            <i
              class="fa fa-lock"
            />
          </div>
        </div>
        <div
          class="p-col-12 row"
        >
          <div
            class="input-wrapper p-float-label"
          >
            <input
              class="p-inputtext p-component"
              data-testid="confirmPassword"
              id="confirmPassword"
              name="confirmPassword"
              type="password"
              value=""
            />
            <label
              for="confirmPassword"
            >
              Confirm Password
            </label>
            <i
              class="fa fa-lock"
            />
          </div>
        </div>
        <div
          class="p-col-12 row center"
        >
          <button
            class="p-button p-component p-button-raised p-button-text-only"
            data-testid="change-password"
            type="button"
          >
            <span
              class="p-button-text p-c"
            >
              Change Password
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
